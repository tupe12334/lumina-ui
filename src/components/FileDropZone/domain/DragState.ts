export type DragStateType = 'idle' | 'dragOver' | 'dragLeave'

export class DragState {
  private constructor(private readonly state: DragStateType) {}

  public static create(state: DragStateType = 'idle'): DragState {
    return new DragState(state)
  }

  public getState(): DragStateType {
    return this.state
  }

  public isIdle(): boolean {
    return this.state === 'idle'
  }

  public isDragOver(): boolean {
    return this.state === 'dragOver'
  }

  public isDragLeave(): boolean {
    return this.state === 'dragLeave'
  }

  public isActive(): boolean {
    return this.state === 'dragOver'
  }

  public setDragOver(): DragState {
    return new DragState('dragOver')
  }

  public setDragLeave(): DragState {
    return new DragState('dragLeave')
  }

  public setIdle(): DragState {
    return new DragState('idle')
  }
}