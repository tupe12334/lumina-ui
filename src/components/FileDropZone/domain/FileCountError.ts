import { FileValidationError } from './FileValidationError'

export class FileCountError extends FileValidationError {
  constructor(actualCount: number, maxCount: number) {
    super(`Cannot upload ${actualCount} files. Maximum allowed: ${maxCount}`)
    this.name = 'FileCountError'
  }
}