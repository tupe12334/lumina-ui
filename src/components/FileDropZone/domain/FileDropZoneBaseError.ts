export class FileDropZoneError extends Error {
  constructor(
    message: string,
    public readonly code: string,
    public readonly fileId?: string
  ) {
    super(message)
    this.name = 'FileDropZoneError'
  }
}