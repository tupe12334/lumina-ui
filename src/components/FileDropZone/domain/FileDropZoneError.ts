export class FileDropZoneError extends Error {
  constructor(
    message: string,
    public readonly code: string,
    public readonly fileId?: string
  ) {
    super(message)
    this.name = 'FileDropZoneError'
  }
}

export class FileValidationError extends FileDropZoneError {
  constructor(message: string, fileId?: string) {
    super(message, 'VALIDATION_ERROR', fileId)
    this.name = 'FileValidationError'
  }
}

export class FileUploadError extends FileDropZoneError {
  constructor(message: string, fileId?: string) {
    super(message, 'UPLOAD_ERROR', fileId)
    this.name = 'FileUploadError'
  }
}

export class FileSizeError extends FileValidationError {
  constructor(actualSize: number, maxSize: number, fileId?: string) {
    const actualSizeMB = Math.round(actualSize / 1024 / 1024)
    const maxSizeMB = Math.round(maxSize / 1024 / 1024)
    super(`File size ${actualSizeMB}MB exceeds maximum allowed size of ${maxSizeMB}MB`, fileId)
    this.name = 'FileSizeError'
  }
}

export class FileTypeError extends FileValidationError {
  constructor(actualType: string, allowedTypes: string[], fileId?: string) {
    const typesList = allowedTypes.join(', ')
    super(`File type "${actualType}" is not allowed. Allowed types: ${typesList}`, fileId)
    this.name = 'FileTypeError'
  }
}

export class FileCountError extends FileValidationError {
  constructor(actualCount: number, maxCount: number) {
    super(`Cannot upload ${actualCount} files. Maximum allowed: ${maxCount}`)
    this.name = 'FileCountError'
  }
}