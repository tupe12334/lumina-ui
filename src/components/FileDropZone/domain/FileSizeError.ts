import { FileValidationError } from './FileValidationError'

export class FileSizeError extends FileValidationError {
  constructor(actualSize: number, maxSize: number, fileId?: string) {
    const actualSizeMB = Math.round(actualSize / 1024 / 1024)
    const maxSizeMB = Math.round(maxSize / 1024 / 1024)
    super(`File size ${actualSizeMB}MB exceeds maximum allowed size of ${maxSizeMB}MB`, fileId)
    this.name = 'FileSizeError'
  }
}