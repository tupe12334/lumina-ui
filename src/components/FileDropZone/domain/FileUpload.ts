import { FileStatusValue, type FileStatus } from './FileStatus'

export interface FileUploadData {
  id: string
  file: File
  status: FileStatus
  progress: number
  error?: string
  previewUrl?: string
}

export class FileUpload {
  private constructor(
    private readonly id: string,
    private readonly file: File,
    private readonly status: FileStatusValue,
    private readonly progress: number,
    private readonly error?: string,
    private readonly previewUrl?: string
  ) {}

  public static create(file: File, id?: string): FileUpload {
    const fileId = id || `${file.name}-${file.size}-${Date.now()}`
    const status = FileStatusValue.create('pending')
    return new FileUpload(fileId, file, status, 0)
  }

  public static fromData(data: FileUploadData): FileUpload {
    const status = FileStatusValue.create(data.status)
    return new FileUpload(
      data.id,
      data.file,
      status,
      data.progress,
      data.error,
      data.previewUrl
    )
  }

  public getId(): string {
    return this.id
  }

  public getFile(): File {
    return this.file
  }

  public getStatus(): FileStatusValue {
    return this.status
  }

  public getProgress(): number {
    return this.progress
  }

  public getError(): string | undefined {
    return this.error
  }

  public getPreviewUrl(): string | undefined {
    return this.previewUrl
  }

  public getName(): string {
    return this.file.name
  }

  public getSize(): number {
    return this.file.size
  }

  public getType(): string {
    return this.file.type
  }

  public withStatus(status: FileStatus): FileUpload {
    const newStatus = FileStatusValue.create(status)
    return new FileUpload(
      this.id,
      this.file,
      newStatus,
      this.progress,
      this.error,
      this.previewUrl
    )
  }

  public withProgress(progress: number): FileUpload {
    const clampedProgress = Math.max(0, Math.min(100, progress))
    return new FileUpload(
      this.id,
      this.file,
      this.status,
      clampedProgress,
      this.error,
      this.previewUrl
    )
  }

  public withError(error: string): FileUpload {
    const failedStatus = FileStatusValue.create('failed')
    return new FileUpload(
      this.id,
      this.file,
      failedStatus,
      this.progress,
      error,
      this.previewUrl
    )
  }

  public withPreviewUrl(previewUrl: string): FileUpload {
    return new FileUpload(
      this.id,
      this.file,
      this.status,
      this.progress,
      this.error,
      previewUrl
    )
  }

  public clearError(): FileUpload {
    return new FileUpload(
      this.id,
      this.file,
      this.status,
      this.progress,
      undefined,
      this.previewUrl
    )
  }

  public toData(): FileUploadData {
    return {
      id: this.id,
      file: this.file,
      status: this.status.getStatus(),
      progress: this.progress,
      error: this.error,
      previewUrl: this.previewUrl
    }
  }
}