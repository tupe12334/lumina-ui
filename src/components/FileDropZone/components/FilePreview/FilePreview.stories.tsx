import type { Meta, StoryObj } from '@storybook/react'
import { FilePreview } from './FilePreview'
import { FileUpload } from '../../domain/FileUpload'

const mockTextFile = new File(['text content'], 'document.txt', { type: 'text/plain' })
const mockPdfFile = new File(['pdf content'], 'document.pdf', { type: 'application/pdf' })
const mockImageFile = new File(['image content'], 'image.jpg', { type: 'image/jpeg' })
const mockVideoFile = new File(['video content'], 'video.mp4', { type: 'video/mp4' })

const meta: Meta<typeof FilePreview> = {
  title: 'Components/FileDropZone/FilePreview',
  component: FilePreview,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
}

export default meta
type Story = StoryObj<typeof meta>

export const WithFileIcon: Story = {
  args: {
    fileUpload: FileUpload.create(mockTextFile, 'text-file'),
  },
}

export const PdfFile: Story = {
  args: {
    fileUpload: FileUpload.create(mockPdfFile, 'pdf-file'),
  },
}

export const VideoFile: Story = {
  args: {
    fileUpload: FileUpload.create(mockVideoFile, 'video-file'),
  },
}

export const WithPreviewImage: Story = {
  args: {
    fileUpload: FileUpload.create(mockImageFile, 'image-file')
      .withPreviewUrl('data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNjQiIGhlaWdodD0iNjQiIHZpZXdCb3g9IjAgMCA2NCA2NCIgZmlsbD0ibm9uZSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48cmVjdCB3aWR0aD0iNjQiIGhlaWdodD0iNjQiIGZpbGw9IiNGRkY2QjQiLz48cGF0aCBkPSJNMjAgMjBMNDQgNDRMMjAgNDRaIiBmaWxsPSIjRkZGRkZGIi8+PC9zdmc+'),
  },
}

export const WithDataURL: Story = {
  args: {
    fileUpload: FileUpload.create(mockImageFile, 'data-url-file')
      .withPreviewUrl('data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNjQiIGhlaWdodD0iNjQiIHZpZXdCb3g9IjAgMCA2NCA2NCIgZmlsbD0ibm9uZSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4KPHJlY3Qgd2lkdGg9IjY0IiBoZWlnaHQ9IjY0IiBmaWxsPSIjRkY2B0I0Ii8+CjxwYXRoIGQ9Ik0yMCAyMEw0NCA0NEwyMCA0NFoiIGZpbGw9IiNGRkZGRkYiLz4KPC9zdmc+'),
  },
}