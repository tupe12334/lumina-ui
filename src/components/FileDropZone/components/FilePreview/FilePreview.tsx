import React from 'react'
import { FileUpload } from '../../domain/FileUpload'
import { FileIcon } from '../../infrastructure/FileIcons'
import styles from '../../FileDropZone.module.css'

interface FilePreviewProps {
  fileUpload: FileUpload
}

export const FilePreview: React.FC<FilePreviewProps> = ({ fileUpload }) => {
  const file = fileUpload.getFile()
  const previewUrl = fileUpload.getPreviewUrl()
  const FileIconComponent = FileIcon

  return (
    <div className={styles.filePreview}>
      {previewUrl ? (
        <img
          src={previewUrl}
          alt={`Preview of ${file.name}`}
          className={styles.previewImage}
        />
      ) : (
        <FileIconComponent className={styles.fileIcon} />
      )}
    </div>
  )
}