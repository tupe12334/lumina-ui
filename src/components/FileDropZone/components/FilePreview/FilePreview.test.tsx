import { render, screen } from '@testing-library/react'
import { FilePreview } from './FilePreview'
import { FileUpload } from '../../domain/FileUpload'

const mockFile = new File(['test content'], 'test.txt', { type: 'text/plain' })

describe('FilePreview', () => {
  test('renders correctly with file icon', () => {
    const fileUpload = FileUpload.create(mockFile, 'test-id')
    const { container } = render(<FilePreview fileUpload={fileUpload} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  test('shows file icon when no preview URL', () => {
    const fileUpload = FileUpload.create(mockFile, 'test-id')
    render(<FilePreview fileUpload={fileUpload} />)

    // Should show file icon (svg element)
    expect(document.querySelector('svg')).toBeInTheDocument()
    // Should not show image
    expect(screen.queryByRole('img')).not.toBeInTheDocument()
  })

  test('shows preview image when preview URL is available', () => {
    const fileUpload = FileUpload.create(mockFile, 'test-id')
      .withPreviewUrl('data:image/png;base64,test')

    render(<FilePreview fileUpload={fileUpload} />)

    const previewImage = screen.getByRole('img')
    expect(previewImage).toBeInTheDocument()
    expect(previewImage).toHaveAttribute('src', 'data:image/png;base64,test')
    expect(previewImage).toHaveAttribute('alt', 'Preview of test.txt')
  })

  test('has correct structure and attributes', () => {
    const fileUpload = FileUpload.create(mockFile, 'test-id')
    const { container } = render(<FilePreview fileUpload={fileUpload} />)

    const previewContainer = container.firstChild
    expect(previewContainer).toBeInTheDocument()
    // Check for CSS module class that contains 'filePreview'
    if (previewContainer) {
      expect(previewContainer.className).toMatch(/filePreview/)
    }
  })

  test('preview image has correct attributes', () => {
    const fileUpload = FileUpload.create(mockFile, 'test-id')
      .withPreviewUrl('data:image/png;base64,test')

    render(<FilePreview fileUpload={fileUpload} />)

    const previewImage = screen.getByRole('img')
    // Check for CSS module class that contains 'previewImage'
    expect(previewImage.className).toMatch(/previewImage/)
  })
})