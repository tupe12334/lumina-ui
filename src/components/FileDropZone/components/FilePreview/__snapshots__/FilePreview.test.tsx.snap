// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FilePreview > renders correctly with file icon 1`] = `
<div
  class="_filePreview_3af3de"
>
  <svg
    class="_fileIcon_3af3de"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
  >
    <path
      class="_fileIcon_3af3de"
      d="M14 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V8z"
    />
    <polyline
      class="_fileIcon_3af3de"
      points="14,2 14,8 20,8"
    />
  </svg>
</div>
`;
